src/dstack/_internal/server/models.py

```
@@ -65,6 +68,59 @@ def process_result_value(self, value, dialect):
         return value
 
 
+class DecryptedString(CoreModel):
+    """
+    A type for representing plaintext strings encrypted with `EncryptedString`.
+    Besides the string, stores information if the decryption was successfull.
+    This is useful so that application code can have custom handling of failed decrypts (e.g. ignoring).
+    """
+
+    plaintext: str
+    decrypted: bool = True
+    exc: Optional[Exception] = None
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class EncryptedString(TypeDecorator):
+    """
+    A custom type decorator that encrypts and decrypts strings for storing in the db.
+    """
+
+    impl = String
+    cache_ok = True
+
+    _encrypt_func: Callable[[str], str]
+    _decrypt_func: Callable[[str], str]
+
+    @classmethod
+    def set_encrypt_decrypt(
+        cls,
+        encrypt_func: Callable[[str], str],
+        decrypt_func: Callable[[str], str],
+    ):
+        cls._encrypt_func = encrypt_func
+        cls._decrypt_func = decrypt_func
+
+    def process_bind_param(self, value: Union[DecryptedString, str], dialect):
+        if isinstance(value, str):
+            # Passing string allows binding an encrypted value directly
+            # e.g. for comparisons
+            return value
+        return EncryptedString._encrypt_func(value.plaintext)
+
+    def process_result_value(self, value: Optional[str], dialect) -> Optional[DecryptedString]:
+        if value is None:
+            return value
+        try:
+            plaintext = EncryptedString._decrypt_func(value)
+            return DecryptedString(plaintext=plaintext, decrypted=True)
+        except Exception as e:
+            logger.debug("Failed to decrypt encrypted string: %s", repr(e))
+            return DecryptedString(plaintext="", decrypted=False, exc=e)
```

It could be subjective, but this design looks a bit error-prone to me. If the caller forgets to check `decrypted`, they may continue using `plaintext=""`, as if decryption worked. This could potentially lead to security issues, such as being able to log in to any account using an empty password.

In other words, in the current design the caller's default is to ignore decryption errors. I think the default should should be to fail.

As an alternative solution, I can suggest decrypting at field access time (i.e. the interface would be `some_field.decrypt()` instead of `some_field.plaintext`) and throwing an exception if decryption fails. And if the caller _really_ needs to ignore the exception, they will catch it and ignore it explicitly.

---

This is a very good point about decryption errors go silent. I wasn't quite satisfied with the access time decryption design I started with so not sure about transitioning to it, but I will definitely replace direct `.plaintext` access with a method that raises an error if there were any.